### PR TITLE
fix(windows): ignore virtual drives

### PR DESCRIPTION
**Pull request by jviotti**
Drivelist currently crashes if there is a virtual drive mounted. The
best solution I could find is to check the resulting error from
IOCTL_STORAGE_GET_DEVICE_NUMBER.

Change-Type: patch
Changelog-Entry: Fix "Couldn't scan the drives: An unknown error occurred" where there is a mounted virtual drive.
Fixes: https://github.com/resin-io-modules/drivelist/issues/198
Signed-off-by: Juan Cruz Viotti <jv@jviotti.com>

### CODE REVIEW

File: src/windows/scanner.cc [hunk 1]
```diff
@@ -186,8 +186,23 @@ ScanMountpoints(std::vector<drivelist::mountpoint_s> *const output) {
...
+    if (FAILED(result)) {
+
+      // This error, "Invalid Function", can happen when attempting
+      // to get the device number of a virtual disk.
+      // See: https://github.com/resin-io-modules/drivelist/issues/198
+      //
+      // TODO: According to the documentation, the ERROR_INVALID_FUNCTION
+      // is the constant we want here, but in practice, that constant
+      // equals "1," and not the hexadecimal number outlined in the docs.
+      // See https://msdn.microsoft.com/en-us/library/windows/desktop/dd319335(v=vs.85).aspx
```
Comment (lurch) [inline src/windows/scanner.cc]: That page specifically relates to "Portable Device" APIs, which we're not using here.
Seems like for the "standard Win32" [APIs](https://msdn.microsoft.com/en-us/library/cc231199.aspx) we're using, `ERROR_INVALID_FUNCTION` does indeed map to `0x1`, and it's just coincidence that `0x80070001` maps to `ERROR_INVALID_FUNCTION` in the "Portable Device" APIs.

I couldn't figure out what constant (if any) `0x80070001` maps to in the "standard Win32" APIs, but I think your talk of "Invalid Function" here (and in the commit message) is a bit of a red herring.
The nearest thing I could find was https://msdn.microsoft.com/en-us/library/cc704587.aspx but that lists nothing between `0x8004E031` and `0x80070005`, so perhaps `0x80070001` is simply an undocumented error code? :confused: 

All AFAICT, of course! :wink:

**Comment by lurch** [comment]
...and the CI builds are all failing because of:
```
> eslint lib tests example && cpplint --recursive src
src/windows/scanner.cc:190:  Redundant blank line at the start of a code block should be deleted.  [whitespace/blank_line] [2]
src/windows/scanner.cc:195:  Missing username in TODO; it should look like "// TODO(my_username): Stuff."  [readability/todo] [2]
```

File: src/windows/scanner.cc (hunk 1 above)
Comment (jviotti) [inline src/windows/scanner.cc]: Oh, I see. Hm, I'll at least update the comment to reflect your findings.

File: src/windows/scanner.cc [hunk 2]
```diff
@@ -186,8 +186,20 @@ ScanMountpoints(std::vector<drivelist::mountpoint_s> *const output) {
     ULONG number;
     drivelist::Debug("Getting device number");
     result = drivelist::volume::GetDeviceNumber(volume, &number);
-    if (FAILED(result))
+    if (FAILED(result)) {
+
+      // This error, "Invalid Function", can happen when attempting
+      // to get the device number of a virtual disk.
```
Comment (lurch) [inline src/windows/scanner.cc]: IMHO you should get rid of the "Invalid Function" part and just change it to `This error can happen when attempting to get the device number of a virtual disk.` with a similar change also made to the commit message.

Comment (jviotti) [inline src/windows/scanner.cc]: Fixed!

**Comment by jviotti** [comment]
Fixed!

**Comment by lurch** [comment]
Does this also fix `SUBST` drives? https://en.wikipedia.org/wiki/SUBST
(or were they not a problem in the first place? Afraid I'm not set up to test on Windows ATM)

**Comment by jviotti** [comment]
It should work fine, because under the hood, there is still just a list of disk volumes.

**Comment by jhermsmeier** [comment]
What exactly qualifies as a "virtual drive"? Mounted VHDs are still (correctly so) listed, for example.

**Comment by jviotti** [comment]
@jhermsmeier Do you know what are all types of virtual drives that we should test?

**Comment by lurch** [comment]
@jhermsmeier I know very little about virtual drives, but see this comment: https://github.com/resin-io-modules/drivelist/issues/198#issuecomment-318591064

**Comment by jhermsmeier** [comment]
Turns out, those aren't virtual drives in that sense, but user-mode file systems – like https://www.eldos.com/cbfs/ and similar. As those do not emulate block devices, they don't have a device number, afaik – hence the error.

**Comment by lurch** [comment]
@jhermsmeier Ahhhh, so does http://www.fs-driver.org/ fall into that same category?

**Comment by jhermsmeier** [comment]
@lurch nope, that's a kernel-mode fs driver (an actual driver), which will be handled by Windows, and has nothing to do with the underlying device.

User-mode file systems (i.e. FUSE, Dokan, Eldos CBFS, etc.) - on the contrary - work by having one common kernel-mode driver that acts as a server to user-space programs that implement the client API to that.

**Pull request merged**